Combine flushes in DnsNameResolver to allow usage of sendmmsg to redu…

normanmaurer opened the pull request:
…ce syscall costs.

Motivation:

Some of transports support gathering writes when using datagrams. For example this is the case for EpollDatagramChannel. We should minimize the calls to flush() to allow making efficient usage of sendmmsg in this case.

Modifications:

- minimize flush() operations by supress these when when we are currently reading from the Channel
- doResolve(...) will typically do multiple queries so supress flush() until we wrote all queries to the Channel

Result:

Be able to use sendmmsg internally in the DnsNameResolver.

ejona86 commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -962,14 +966,45 @@ static boolean doResolveAllCached(String hostname,
-                                     resolveCache, authoritativeDnsServerCache).resolve(promise);
+
+        try {
+            // As the resolve(...) method may trigger multiple flush() operations lets just suppress these until we
```
Where are these flushes being suppressed coming from? Why can't the be removed?

ejona86 commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -962,14 +966,45 @@ static boolean doResolveAllCached(String hostname,
+        try {
+            // As the resolve(...) method may trigger multiple flush() operations lets just suppress these until we
+            // are done with the resolve(...) operation and then flush all together to reduce syscalls.
+            handler.supressFlushes();
```
Is reentrancy a problem? It looks like it would be possible.

johnou commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -1100,16 +1135,22 @@ final DnsServerAddressStream newNameServerAddressStream(String hostname) {
 
         private final Promise<Channel> channelActivePromise;
+        private ChannelHandlerContext ctx;
+        private boolean supressFlushes;
```
supress => suppress

normanmaurer commented:
@ejona86 @carl-mastrangelo @johnou PTAL again... I reworked this one to be more clear...

normanmaurer commented:
@netty-bot test this please.

carl-mastrangelo commented:
@normanmaurer FYI, it would be helpful for code review if you squashed just before merging the PR.  GitHub's UI can show changes since last review, but not if the commits are squashed.

normanmaurer commented:
Ah yeah sorry I completely rewrote it so I thought it makes more sense to force push here

> Am 16.11.2018 um 19:51 schrieb Carl Mastrangelo <notifications@github.com>:
> 
> @normanmaurer FYI, it would be helpful for code review if you squashed just before merging the PR. GitHub's UI can show changes since last review, but not if the commits are squashed.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

normanmaurer commented:
@ejona86 does this look better now ?

normanmaurer commented:
@ejona86 yep.
